### PR TITLE
Update hardware interface for ROS2

### PR DESCRIPTION
**Pull request by livanov93**
The idea is to implement UR position hardware interface for ROS2 by inheriting [ros2_control base class](https://github.com/ros-controls/ros2_control/blob/master/hardware_interface/include/hardware_interface/components/system_interface.hpp).
The analogy with ROS1 would be [RobotHardware class](https://github.com/ros-controls/ros_control/blob/noetic-devel/hardware_interface/include/hardware_interface/robot_hw.h) and [UR hardware interface](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/blob/master/ur_robot_driver/include/ur_robot_driver/hardware_interface.h).

### CODE REVIEW

File: ur_robot_driver/include/ur_robot_driver/hardware_interface.h [hunk 1]
```diff
@@ -55,26 +59,94 @@ namespace ur_robot_driver
...
+class URPositionHardwareInterface : public hardware_interface::SystemInterface
 {
 public:
-  RCLCPP_SHARED_PTR_DEFINITIONS(URHardwareInterface);
+  RCLCPP_SHARED_PTR_DEFINITIONS(URPositionHardwareInterface);
+
+  return_type configure(const HardwareInfo& system_info) final;
 
-  hardware_interface::return_type configure(const HardwareInfo& system_info) final;
+  std::vector<hardware_interface::StateInterface> export_state_interfaces() final;
```
Comment (AndyZe) [inline ur_robot_driver/include/ur_robot_driver/hardware_interface.h:69]: member functions should be camelCased. e.g. `exportStartInterfaces`, `getStatus`

Comment (AndyZe) [reply ur_robot_driver/include/ur_robot_driver/hardware_interface.h:69]: Oh, these were inherited, huh? I guess the naming issue is on the ros2_control side, then

Comment (livanov93) [reply ur_robot_driver/include/ur_robot_driver/hardware_interface.h:69]: Yes , it is inherited

Comment (AndyZe) [reply ur_robot_driver/include/ur_robot_driver/hardware_interface.h:69]: Made an issue here we can keep an eye on

https://github.com/ros-controls/ros2_control/issues/269

**Review by AndyZe — COMMENTED** [review]
Overall, looks good. I'll have to closely review `dashboard_client_node.cpp` later, I ran out of time for now.

Does it make sense to move `ur_robot_driver/resources` into `ur5_moveit_config`? Since it is setup stuff not related to the driver. The only problem I see with that is we would end up with a copy for each robot type. Maybe we should start a whole new package for demo launch files.

**Comment by AndyZe** [comment]
@gavanderhoorn do you have any thought on this:

> Does it make sense to move ur_robot_driver/resources into ur5_moveit_config? Since it is setup stuff not related to the driver. The only problem I see with that is we would end up with a copy for each robot type. Maybe we should start a whole new package for demo launch files.

File: ur_robot_driver/src/hardware_interface.cpp [hunk 2]
```diff
@@ -25,26 +25,429 @@
...
+  if (data_pkg)
+  {
+    packet_read_ = true;
+    readData(data_pkg, "actual_q", urcl_joint_positions_);
+    readData(data_pkg, "actual_qd", urcl_joint_velocities_);
+    readData(data_pkg, "actual_current", urcl_joint_efforts_);
+
+    memcpy(&position_states_[0], &urcl_joint_positions_[0], 6 * sizeof(double));
+    memcpy(&velocity_states_[0], &urcl_joint_velocities_[0], 6 * sizeof(double));
+    memcpy(&joint_efforts_[0], &urcl_joint_efforts_[0], 6 * sizeof(double));
```
Comment (AndyZe) [inline ur_robot_driver/src/hardware_interface.cpp:352]: memcpy makes me a little nervous. I just don't see it used very often.

Why is it needed here but not in the [ROS1 version](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/blob/master/ur_robot_driver/src/hardware_interface.cpp#L430)?

File: ur_robot_driver/src/hardware_interface.cpp [hunk 3]
```diff
@@ -25,26 +25,429 @@
...
+    else
+    {
+      ur_driver_->writeKeepalive();
+    }
+
+    packet_read_ = false;
+
+    // remember old values
+    memcpy(&position_commands_old_[0], &position_commands_[0], 6 * sizeof(double));
+    memcpy(&velocity_commands_old_[0], &velocity_commands_[0], 6 * sizeof(double));
```
Comment (AndyZe) [inline ur_robot_driver/src/hardware_interface.cpp:434]: Why not simply do `velocity_commands_old_ = velocity_commands_`?

File: ur_robot_driver/src/hardware_interface.cpp [hunk 4]
```diff
@@ -25,26 +25,429 @@
...
+
+    return return_type::OK;
+  }
+  else
+  {
+    // TODO could not read form the driver --> reset controllers
```
Comment (AndyZe) [inline ur_robot_driver/src/hardware_interface.cpp:440]: form -> from

**Comment by livanov93** [comment]
> Overall, looks good. I'll have to closely review `dashboard_client_node.cpp` later, I ran out of time for now.
> 
> Does it make sense to move `ur_robot_driver/resources` into `ur5_moveit_config`? Since it is setup stuff not related to the driver. The only problem I see with that is we would end up with a copy for each robot type. Maybe we should start a whole new package for demo launch files.

I planned to remove it at the point when the whole package is functional. It is only used for testing purposes.
I will move it to the demo subpackage within the Univeral_Robots_ROS2_Driver. @AndyZe Do you agree?

File: ur_robot_driver/resources/servo/start_configuration [hunk 5]
```diff
@@ -0,0 +1,7 @@
+ros2 topic pub --rate 2 /forward_command_controller_position/commands std_msgs/msg/Float64MultiArray "data:
+- 0.5
+- -2.0
+- 2.0
+- 0.0
+- 1.57
+- 0.0"
```
[the remark this answers is not in the document]
Comment (livanov93) [reply ur_robot_driver/resources/servo/start_configuration:7]: It is unneeded. Will be removed

File: ur_robot_driver/src/hardware_interface.cpp (hunk 2 above)
Comment (livanov93) [reply ur_robot_driver/src/hardware_interface.cpp:352]: I will use the `urcl_<>` data storage everywhere.

File: ur_robot_driver/src/hardware_interface.cpp [hunk 6]
```diff
@@ -25,26 +25,429 @@
...
+  // A higher gain can sharpen the trajectory.
+  int servoj_gain = stoi(info_.hardware_parameters["servoj_gain"]);
+  // Specify lookahead time for servoing to position in joint space.
+  // A longer lookahead time can smooth the trajectory.
+  double servoj_lookahead_time = stod(info_.hardware_parameters["servoj_lookahead_time"]);
+
+  // Whenever the runtime state of the "External Control" program node in the UR-program changes, a
+  // message gets published here. So this is equivalent to the information whether the robot accepts
+  // commands from ROS side.
+  //  program_state_pub_ = robot_hw_nh.advertise<std_msgs::Bool>("robot_program_running", 10, true);
```
[the remark this answers is not in the document]
Comment (livanov93) [reply ur_robot_driver/src/hardware_interface.cpp:195]: Accidentally left

File: ur_robot_driver/src/hardware_interface.cpp [hunk 7]
```diff
@@ -25,26 +25,429 @@
...
+  // Path to the urscript code that will be sent to the robot
+  std::string script_filename = info_.hardware_parameters["script_filename"];
+  // Path to the file containing the recipe used for requesting RTDE outputs.
+  std::string output_recipe_filename = info_.hardware_parameters["output_recipe_filename"];
+  // Path to the file containing the recipe used for requesting RTDE inputs.
+  std::string input_recipe_filename = info_.hardware_parameters["input_recipe_filename"];
+  // Start robot in headless mode. This does not require the 'External Control' URCap to be running
+  // on the robot, but this will send the URScript to the robot directly. On e-Series robots this
+  // requires the robot to run in 'remote-control' mode.
+  bool headless_mode = static_cast<bool>(stoi(info_.hardware_parameters["headless_mode"]));
```
[the remark this answers is not in the document]
Comment (livanov93) [reply ur_robot_driver/src/hardware_interface.cpp:166]: I had problems in adding the boolean parameter in the ros2_control.xacro file. Parsing was interrupted and the ros2_control_node was terminated. 
This is a workaround!

[the remark this answers is not in the document]
Comment (livanov93) [reply ur_robot_driver/src/hardware_interface.cpp:166]: Could be changed with [`boost::lexical_cast<bool>(info_.hardware_parameters["headless_mode"])`](https://www.boost.org/doc/libs/1_36_0/libs/conversion/lexical_cast.htm#lexical_cast).

File: ur_robot_driver/CMakeLists.txt [hunk 8]
```diff
@@ -67,6 +67,9 @@ add_executable(dashboard_client
 )
 target_link_libraries(dashboard_client ${catkin_LIBRARIES} ur_client_library::urcl)
 ament_target_dependencies(dashboard_client ${${PROJECT_NAME}_EXPORTED_TARGETS} ${THIS_PACKAGE_INCLUDE_DEPENDS})
+install(TARGETS
+        dashboard_client
```
[the remark this answers is not in the document]
Comment (AndyZe) [reply ur_robot_driver/CMakeLists.txt:72]: I mean, it should be on the same line as Targets. Like this:

install(
  TARGETS dashboard_client
  DESTINATION ...
)

To match the format of the other commands here

File: ur_robot_driver/launch/ur_dashboard_client.launch.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+            package="ur_robot_driver",
+            executable="dashboard_client",
+            name="dashboard_client",
+            output="screen",
+            emulate_tty=True,
+            parameters=[
+                {"robot_ip": "000.000.000.000"}
+            ]
+        )
+    ])
```
Comment (AndyZe) [inline ur_robot_driver/launch/ur_dashboard_client.launch.py:16]: nitpick: leave a blank line at the end of the file

File: ur_robot_driver/src/hardware_interface.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (AndyZe) [reply ur_robot_driver/src/hardware_interface.cpp:166]: I guess it's fine for now. As long as you had a good reason to do this

File: ur_ros2_control_demos/config/ur5_simulated_config.yaml [hunk 10]
```diff
@@ -0,0 +1,68 @@
+###############################################
```
Comment (AndyZe) [inline ur_ros2_control_demos/config/ur5_simulated_config.yaml:1]: I think this folder is an OK place to stick demo stuff for now. We'll clean it up later

Comment (AndyZe) [reply ur_ros2_control_demos/config/ur5_simulated_config.yaml:1]: Please rename this file:  `ur5_servo_config.yaml`

File: ur_ros2_control_demos/urdf/ur5.urdf [hunk 11]
```diff
@@ -0,0 +1,241 @@
+<?xml version="1.0" ?>
```
Comment (AndyZe) [inline ur_ros2_control_demos/urdf/ur5.urdf:1]: Please create an issue to delete this duplicate urdf (or just delete it now). It's a duplicate of the one in ur_description/urdf

File: ur_ros2_control_demos/urdf/ur5.urdf.xacro [hunk 12]
```diff
@@ -0,0 +1,209 @@
+?xml version="1.0"?>
```
Comment (AndyZe) [inline ur_ros2_control_demos/urdf/ur5.urdf.xacro:1]: Same comment about this urdf being a duplicate of the one in ur_description/urdf

File: ur_ros2_control_demos/urdf/ur5_system_position_only.ros2_control.xacro [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        <command_interface name="velocity">
+          <param name="min">-3.2</param>
+          <param name="max">3.2</param>
+        </command_interface>
+        <state_interface name="position"/>
+        <state_interface name="velocity"/>
+        <state_interface name="effort"/>
+      </joint>
+    </ros2_control>
+
```
Comment (AndyZe) [inline ur_ros2_control_demos/urdf/ur5_system_position_only.ros2_control.xacro:110]: delete this empty line

File: ur_ros2_control_demos/urdf/ur5_system_position_only.ros2_control.xacro [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+  </xacro:macro>
+
```
Comment (AndyZe) [inline ur_ros2_control_demos/urdf/ur5_system_position_only.ros2_control.xacro:112]: delete this empty line

**Review by AndyZe — COMMENTED** [review]
@livanov93 Just a couple minor comments left to address, then it looks ready to merge.  :+1:

File: ur_ros2_control_demos/urdf/ur5_system_position_only.ros2_control.xacro (hunk 14 above)
Comment (livanov93) [reply ur_ros2_control_demos/urdf/ur5_system_position_only.ros2_control.xacro:112]: done

File: ur_ros2_control_demos/urdf/ur5.urdf (hunk 11 above)
Comment (livanov93) [reply ur_ros2_control_demos/urdf/ur5.urdf:1]: It is not the same, it has included ros2_control macro. Will hold it for know

File: ur_ros2_control_demos/urdf/ur5.urdf.xacro (hunk 12 above)
Comment (livanov93) [reply ur_ros2_control_demos/urdf/ur5.urdf.xacro:1]: It is not the same, it has included ros2_control macro. Will hold it for know

File: ur_ros2_control_demos/config/ur5_simulated_config.yaml (hunk 10 above)
Comment (livanov93) [reply ur_ros2_control_demos/config/ur5_simulated_config.yaml:1]: will do

**Review by AndyZe — APPROVED** [review]

**Comment by AndyZe** [comment]
There are a couple whitespace cleanup things to clean up still, but I can do that in a followup PR.

**Pull request merged**